data/transactions/logic/frames.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"errors"
+	"fmt"
+)
+
+func opProto(cx *EvalContext) error {
+	if !cx.fromCallsub {
+		return fmt.Errorf("proto executed in normal flow")
+	}
+	cx.fromCallsub = false
+	nargs := int(cx.program[cx.pc+1])
+	if nargs > len(cx.stack) {
+		return fmt.Errorf("callsub to proto that requires %d args with stack height %d", nargs, len(cx.stack))
+	}
+	top := len(cx.callstack) - 1
+	cx.callstack[top].clear = true
+	cx.callstack[top].args = nargs
+	cx.callstack[top].returns = int(cx.program[cx.pc+2])
+	return nil
+}
+
+func opFrameDig(cx *EvalContext) error {
+	i := int8(cx.program[cx.pc+1])
```

Comment:
I'm on the fence about whether `frame_dig` should be allowed to go *deeper* than args. My gut feeling is that if you used `proto`, it should not be allowed, but if you didn't, you might be implementing a varargs style subroutine, and you might need to go arbitrarily deep.

You should (definitely* be allowed to go *above* the return values though.  That is where you would put local variables.

Reply:
One use case that strikes me as commonplace would be to do something like split a `byte[]` onto the stack as individual stack elements and then process. That's where the varargs approach would useful. On the other hand, the split function itself has the opposite problem - you can't hardcode the number of returns. To formalize this one could introduce:

* `proto * n`
* `proto m *`
* `proto * *`

Reply:
Since `proto` is not required, my inclination is to treat a lack of `proto` as what I think you mean by `proto * *`.

I think `proto m *` is a very weird case. Unknown number of return values?  How would the caller process them?  It also eliminates one of the main advantages of using `proto` - it cleans up the stack above the return values for you.

`proto * n` has a nice interpretation, but again makes it impossible for `proto` to do it's job - cause callsub to slide the returns down over the args, and clean up the stack.

Reply:
Agree that `*` is a can of worms that would require too many changes and not worth pursuing at this juncture.